### ISSUE TITLE
Incorrectly generated applications in Application.spec/1 for 1.7.x

### ISSUE THREAD

**Issue by midas** [elixir-lang/elixir#8733]
### Environment
Erlang/OTP 21 [erts-10.2.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]
Elixir 1.7.4 (compiled with OTP 19)
Operating system: Fedora 29

### Current behavior

It appears the `applications` in `Application.spec/1` is being incorrectly calculated.

Given an application, `:backend_nm` with mix deps defined as such:
```elixir
  defp deps do
    [
      {:state_switch_mock, in_umbrella: true, only: ~w(dev integration test)a}, # necessary for order of startup

      {:backend, in_umbrella: true},
      {:parser, in_umbrella: true},
      {:service, in_umbrella: true},
      {:shared, in_umbrella: true},
      {:test_support, in_umbrella: true, only: ~w(integration test)a},

      {:nodefinder, github: "erlanger/nodefinder", manager: :rebar}
    ]
  end
```
with a `MIX_ENV` of `prod` `Application.spec(:backend_nm)` is returning:

```elixir
[
  description: 'backend_nm',
  ...
  included_applications: [],
  applications: [:kernel, :stdlib, :elixir, :logger, :nodefinder,
   :test_support, :shared, :parser, :service, :backend, :state_switch_mock],
  ...
]
```

Notice the `:test_support` and `:state_switch_mock` apps are incorrectly included as they are defined for `~w(integration test)a` and `~w(dev integration test)a` respectively 

The same results are reached with Elixir versions `1.7.0`, `1.7.1`, `1.7.2`, `1.7.3` and `1.7.4`.

### Expected behavior

Given the same deps defined for application `:backend_nm`, Elixir 1.6.6
with a `MIX_ENV` of `prod` `Application.spec(:backend_nm)` is correctly returning:
```elixir
[
  description: 'backend_nm',
  ...
  included_applications: [],
  applications: [:kernel, :stdlib, :elixir, :logger, :nodefinder,
   :shared, :parser, :service, :backend],
  ...
]
```

**Comment by josevalim** [comment]
Hi @midas, can you please provide a simple application that  reproduces this? It will make our lives much easier to reproduce and fix it. Thanks!

**Comment by josevalim** [comment]
To clarify, I could not reproduce this:

```
$ mix new umb --umbrella
$ cd umb/apps
$ mix new foo
$ mix new bar
$ # add {:bar, in_umbrella: true, only: [:dev, :test]} to foo/mix.exs
$ cd foo
$ MIX_ENV=prod mix run -e "IO.inspect Application.spec :foo, :applications"
Compiling 1 file (.ex)
Generated foo app
[:kernel, :stdlib, :elixir, :logger]
```

**Comment by midas** [comment]
@josevalim Hi Jose.  Thanks for the response.  Here is a project I generated exactly as you did: https://github.com/midas/umb.

I did not see anything in your response about the Elixir version you used.  After following your exact process detailed above I got these incorrect result.  By the way, I am using `asdf` to switch Elixir versions.

```
cjharrelson@host:~/sources/personal/umb » elixir --version
Erlang/OTP 20 [erts-9.1.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]

Elixir 1.7.4 (compiled with Erlang/OTP 19)

cjharrelson@host:~/sources/personal/umb » MIX_ENV=prod mix run -e "IO.inspect Application.spec :foo, :applications"
==> bar
Compiling 1 file (.ex)
Generated bar app
==> foo
Compiling 1 file (.ex)
Generated foo app
[:kernel, :stdlib, :elixir, :logger, :bar]
```

Here it is functioning correctly with Elixir 1.6.6:
```
cjharrelson@host:~/sources/personal/umb » rm -rf _build

cjharrelson@host:~/sources/personal/umb » elixir --version
Erlang/OTP 20 [erts-9.1.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]

Elixir 1.6.6 (compiled with OTP 19)

cjharrelson@host:~/sources/personal/umb » MIX_ENV=prod mix run -e "IO.inspect Application.spec :foo, :applications"  
==> bar
Compiling 1 file (.ex)
Generated bar app
==> foo
Compiling 1 file (.ex)
Generated foo app
[:kernel, :stdlib, :elixir, :logger]
```

Here it is functioning correctly with Elixir 1.8.0:
```
cjharrelson@host:~/sources/personal/umb » rm -rf _build  

cjharrelson@host:~/sources/personal/umb » elixir --version 
Erlang/OTP 20 [erts-9.1.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]

Elixir 1.8.0 (compiled with Erlang/OTP 20)

cjharrelson@host~/sources/personal/umb » MIX_ENV=prod mix run -e "IO.inspect Application.spec :foo, :applications"
==> bar
Compiling 1 file (.ex)
Generated bar app
==> foo
Compiling 1 file (.ex)
Generated foo app
[:kernel, :stdlib, :elixir, :logger]
```

I noticed if I switch Elixir versions and do not remove the `_build` directory prior to comilation the project is apparently not fully recompiling and will act like the last fully recompiled version.  Thus if I start on Elixir `1.7.4` and switch to `1.6.6` without removing `_build`, it still produces the incorrect result.  Conversey, if I compile for the first time on Elixir `1.6.6` and switch to `1.7.x` I get the correct results.

Thank you in advance for your concern.  I appreciate all that you do for us!

**Comment by josevalim** [comment]
> I noticed if I switch Elixir versions and do not remove the _build directory prior to comilation the project is apparently not fully recompiling and will act like the last fully recompiled version. 

Ah, good catch, that's why I could not reproduce it on v1.7.x the first time around. However, as you mentioned, the issue is fixed on v1.8.0, so we recommend developers to migrate to the latest. v1.7 is on security bug fixes only.

Thanks for the report!
